active/0002-centalized-mgmt-resource-types-provider-versions.md

```diff
@@ -0,0 +1,277 @@
+    }
+}
+```
+
+- `src/modules/mod.bicep`
+
+```bicep
+provider bar as modBarProvider
+provider foo as modFooProvider
+```
+
+- `src/modules/bicepconfig.json`
+
+```bicep
+{
+     "experimentalFeaturesEnabled": {
+        "compileTimeImports": false
+     },
+}
+```
+
+The user may expect Bicep to resolve the provider details of `modFooProvider` by inspection of `src/bicepconfig.json` and for this project to compile successfully in a way that feels familiar with other programming languages where merge semantics are applied on code dependencies, however, Bicep config merge semantics uses the configuration file closest to the Bicep file in the directory hierarchy and merges them to the builtin config. It doesn't merge other configs in the file hierarchy prior to merging with the builtin config.
+
+In this case Bicep will try to resolve the `modFooProvider` details from `src/modules/bicepconfig.json`. The config doesn't specify a `providers` section, and the builtin config doesn't include a definition for provider `foo` and so the compiler cannot resolve the provider declaration statement and will emit a diagnostic.
+
+If we replace `foo` with `az` in the example above, Bicep resolves `modFooProvider`'s version to be the version of `az` specified in the builtin configuration and not the version `1.2.3` specified in `src/modules/mod.bicep` (inherited from `src/bicepconfig.json`) which may also be unexpected. The reson for this behavior is that current `bicepconfig.json` merge semantics merge `src/modules/bicepconfig.json` with the builtin config and the version applied would be the value specified in the builtin az provider version.
+
+This is especially problematic for 3rd party namespace identifiers specified in a `bicepconfig.json` up a file hierarchy, since they would result in a compilation error because they are unspecified in the builtin configuration.
+
+### Resolution of namespace identifiers by VSCode extension when inspecting published module sources
+
+The ability to inspect sources is being implemented in Bicep to allow users to see what sources were used to compile a Bicep module, since the version information is codified in the `bicepconfig.json` and that is not published with the sources, the costumer will not be able to immediately know what version of a provider was used to resolve the resources in the souces file. This is illustrated in the example below.
+
+**Example: User consumes a published module (that uses new provider syntax) and inspects its sources**
+
+main.bicep
+
+```bicep
+provider az
+
+module myMod 'br:mcr.microsoft.com/bicep/path/to/my_mod:1.2.3' = {
+  name: 'example'
+  params: {}
+}
+```
+
+Let's assume that `my_mod` was published from a Bicep project with the following `bicepconfig.json`:
+
+bicepconfig.json
+
+```jsonc
+{
+  "providers": {
+    "az": {
+      "source": {
+        "registry": "mcr.microsoft.com/bicep/providers",
+        "version": "0.2.3"
+      },
+      "isDefaultImport": true
+    }
+  }
```

Comment:
Should we consider how to handle conflicts? Two scenarios come to mind:

- The user tries to import `az` from both MCR and ACR.
- The user tries to import a third-party provider whose name happens to be `az` 😈. Although we don't support third-party providers today, it might be worth thinking through the design.

Reply:
Thanks for these comments, they are on point and useful: 
On the first point I made a change to fully qualify the repository in the `sources` key, also, added a constraint so that all keys in the `providers` object are unique. I believe this solves the issue as the user is now able to import both provider packages and name their identifier without name collision.
The second scenario seems to be an instance of the same limitation that was addressed by the changes mentioned above.